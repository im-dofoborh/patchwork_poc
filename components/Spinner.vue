<template>
  <div class="sg-spinner">
    <transition name="fade">
      <div v-if="show">
        <letter-icon
          :size="letterSize"
          :colour="color"
          word="Patchwork"
          class="sg-spinner__letter" />
      </div>
    </transition>
  </div>
</template>

<script>
  import LetterIcon from '~/components/icons/Letter'

  export default {
    components: {
      LetterIcon
    },

    props: {
      size: {
        type: String,
        default: 'md',
      },
      light: {
        type: Boolean,
        default: false,
      },
    },

    data() {
      return {
        show: false,
      }
    },

    computed: {
      letterSize() {
        const sizes = {
          sm: 25,
          md: 50,
          lg: 75,
        }
        return sizes[this.size] || sizes.sm;
      },

      color() {
        return this.light ? 'rgba(255, 255, 255, 0.15)' : 'rgba(0, 0, 0, 0.2)';
      }
    },

    mounted() {
      this.show = true; // Triggers transition
    }
  }
</script>

<style lang="scss" scoped>
  .sg-spinner {
    &__letter {
      animation: spinner 2s infinite linear;
    }

    @keyframes spinner {
      0% {
        transform: perspective(120px) rotateX(0deg) rotateY(0deg);
      }
      25% {
        transform: perspective(120px) rotateX(-180.1deg) rotateY(0deg);
      }
      50% {
        transform: perspective(120px) rotateX(-180deg) rotateY(-179.9deg);
      }
      75% {
        transform: perspective(120px) rotateX(0.1deg) rotateY(-180.1deg);
      }
    }

    .fade-enter-active, .fade-leave-active {
      transition: opacity 5s;
    }

    .fade-enter, .fade-leave-to {
      opacity: 0;
    }
  }
</style>
