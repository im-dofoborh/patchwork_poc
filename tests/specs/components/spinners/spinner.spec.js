import { mount, createLocalVue } from '@vue/test-utils';
import Spinner from '~/components/Spinner';

describe('Spinner component', () => {
  let localVue = null;
  let wrapper = null;

  beforeEach(() => {
    localVue = createLocalVue();
    localVue.component('spinner', Spinner);
    wrapper = mount(Spinner, {
      localVue,
    })
  })

  it('renders default variant', () => {
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('renders small variant', () => {
    wrapper.setProps({
      size: 'sm',
    })
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('renders large variant', () => {
    wrapper.setProps({
      size: 'lg',
    })
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('renders medium variant as fallback', () => {
    wrapper.setProps({
      size: 'some-invalid-size',
    })
    expect(wrapper.html()).toMatchSnapshot();
  });

  it('renders light variant', () => {
    wrapper.setProps({
      light: true,
    })
    expect(wrapper.html()).toMatchSnapshot();
  });
});
