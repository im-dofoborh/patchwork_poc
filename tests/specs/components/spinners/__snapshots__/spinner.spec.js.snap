// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spinner component renders default variant 1`] = `
<div class="sg-spinner">
  <div name="fade">
    <div title="Patchwork" class="sg-letter-icon sg-spinner__letter" style="width: 50px;">
      <div class="sg-letter-icon__inner" style="height: 50px; width: 50px; background-color: rgba(0, 0, 0, 0.2); color: rgb(255, 255, 255); font-size: 27.77777777777778px;">
        P
      </div>
    </div>
  </div>
</div>
`;

exports[`Spinner component renders large variant 1`] = `
<div class="sg-spinner">
  <div name="fade">
    <div title="Patchwork" class="sg-letter-icon sg-spinner__letter" style="width: 50px;">
      <div class="sg-letter-icon__inner" style="height: 50px; width: 50px; background-color: rgba(0, 0, 0, 0.2); color: rgb(255, 255, 255); font-size: 27.77777777777778px;">
        P
      </div>
    </div>
  </div>
</div>
`;

exports[`Spinner component renders light variant 1`] = `
<div class="sg-spinner">
  <div name="fade">
    <div title="Patchwork" class="sg-letter-icon sg-spinner__letter" style="width: 50px;">
      <div class="sg-letter-icon__inner" style="height: 50px; width: 50px; background-color: rgba(0, 0, 0, 0.2); color: rgb(255, 255, 255); font-size: 27.77777777777778px;">
        P
      </div>
    </div>
  </div>
</div>
`;

exports[`Spinner component renders medium variant as fallback 1`] = `
<div class="sg-spinner">
  <div name="fade">
    <div title="Patchwork" class="sg-letter-icon sg-spinner__letter" style="width: 50px;">
      <div class="sg-letter-icon__inner" style="height: 50px; width: 50px; background-color: rgba(0, 0, 0, 0.2); color: rgb(255, 255, 255); font-size: 27.77777777777778px;">
        P
      </div>
    </div>
  </div>
</div>
`;

exports[`Spinner component renders small variant 1`] = `
<div class="sg-spinner">
  <div name="fade">
    <div title="Patchwork" class="sg-letter-icon sg-spinner__letter" style="width: 50px;">
      <div class="sg-letter-icon__inner" style="height: 50px; width: 50px; background-color: rgba(0, 0, 0, 0.2); color: rgb(255, 255, 255); font-size: 27.77777777777778px;">
        P
      </div>
    </div>
  </div>
</div>
`;
